src/dune_engine/scheduler.ml

```diff
@@ -1237,6 +1239,23 @@ module Run = struct
     let initial_invalidation = Fs_memo.init ~dune_file_watcher:file_watcher in
     Memo.reset initial_invalidation;
     let result =
+      let run =
+        match timeout with
+        | None -> run
+        | Some timeout ->
+          fun () ->
+            let sleep = Alarm_clock.sleep (Lazy.force t.alarm_clock) timeout in
+            Fiber.fork_and_join_unit
+              (fun () ->
+                let+ res = Alarm_clock.await sleep in
+                match res with
+                | `Finished -> Event_queue.send_signal t.events Shutdown
+                | `Cancelled -> ())
+              (fun () ->
+                Fiber.finalize run ~finally:(fun () ->
+                    Alarm_clock.cancel (Lazy.force t.alarm_clock) sleep;
+                    Fiber.return ()))
+      in
```

Comment:
@aalekseyev this is worth a look as well. It's pretty useful to debug tests.

Reply:
This looks good to me.